[Startup Independence] Part 2: Failing Tests Fixes

gmaretic opened the pull request:
**Goals (and why)**:
Fixed missing initialization and removed the initialization check from the method that is supposed to be idempotent.

**Implementation Description (bullets)**:
Pretty straightforward

**Concerns (what feedback would you like?)**:
The table creation that happens in TransactionManagers.create is idempotent. However, the state can change, as we can actually delete some of the tables in between two calls to TransactionManagers.create. Therefore, a second call to TM.create has to attempt to recreate tables regardless of whether it has been initialized or not.
Need to check that the initialization check is correct in all other classes that use async initialization, but this is out of the scope of this fix. 

**Where should we start reviewing?**:
Wherever

**Priority (whenever / two weeks / yesterday)**:
Soonish

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/palantir/atlasdb/2300)
<!-- Reviewable:end -->

gmaretic commented:
This is completely fixed in https://github.com/palantir/atlasdb/pull/2308, and also blocked on that PR first.

jeremyk-91 commented on atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/MultiCassandraStartupOrderingEteTest.java:
```diff
@@ -138,22 +128,22 @@ private static void assertNotSatisfiedWithin(long time, Callable<Boolean> condit
+            return true;
+        } catch (Exception e) {
             return false;
         }
```
Since `canPerformTransaction()` catches `AtlasDbRemoteException` and just returns false if that's the case, we don't need the `catch AtlasDbRemoteException` block here.

(I realise this is an issue with the original impl though!)

jeremyk-91 commented on atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/MultiCassandraStartupOrderingEteTest.java:
```diff
@@ -68,56 +69,45 @@ public void randomizeKeyspace() throws IOException, InterruptedException {
+        assertSatisfiedWithin(10, () -> !serverRunning());
     }
 
     private static void startTheAtlasServer() throws IOException, InterruptedException {
```
nit: For consistency with the stop-method I think we should name this `startAtlasServer()`.

jeremyk-91 commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/LockStoreImpl.java:
```diff
@@ -82,7 +81,7 @@ public LockStore delegate() {
     private static final Logger log = LoggerFactory.getLogger(LockStoreImpl.class);
     private static final String BACKUP_LOCK_NAME = "BackupLock";
-    private static final AtomicBoolean isInitialized = new AtomicBoolean(false);
+    private volatile boolean isInitialized = false;
```
Hmm. There seems to be a change in semantics here; if I create a transaction manager TM1, close it, and then create TM2 (all in the same JVM), then we previously fail but now we succeed (which makes sense). I guess that makes sense.

fsamuel-bs replied:
Agreed that the current logic makes more sense. We should probably check that this won't break the internal large product though. Maybe running the tests with a local patch would be enough?

jeremyk-91 commented on atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/MultiCassandraStartupOrderingEteTest.java:
```diff
@@ -68,56 +69,45 @@ public void randomizeKeyspace() throws IOException, InterruptedException {
+    @After
+    public void restoreCassandra() throws InterruptedException {
+        startCassandraNodes();
     }
```
It seems unexpected that we wouldn't need to reset the keyspace, though I guess the sed runs within the docker container and we don't reuse ETE-server containers.

fsamuel-bs commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/CellsSweeper.java:
```diff
@@ -55,15 +55,23 @@ public CellsSweeper(
+    }
+
+    public CellsSweeper(
+            TransactionManager txManager,
```
The indentation between this function and the above function is different. This will probably break checkstyle.

fsamuel-bs commented on atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/MultiCassandraStartupOrderingEteTest.java:
```diff
@@ -68,56 +69,45 @@ public void randomizeKeyspace() throws IOException, InterruptedException {
-
-        stopTheAtlasServer();
+        restartAtlasWithDegradedCassandra();
+        Assert.assertEquals(hasQuorum(), canPerformTransaction());
```
Not sure what we're testing with this line. Maybe this changed in subsequent PRs?

fsamuel-bs commented on atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraVerifier.java:
```diff
@@ -211,7 +211,8 @@ private static void attemptToCreateKeyspaceOnHost(InetSocketAddress host, Cassan
                 "(adding the initial empty keyspace)",
-                config.schemaMutationTimeoutMillis());
+                config.schemaMutationTimeoutMillis(),
+                true);
```
This is needed to allow our ETE tests to run with 2 C* nodes up on a 3 node C* cluster. Partially fixes https://github.com/palantir/atlasdb/issues/1705.

gsheasby commented:
Reviewed 1 of 10 files at r1, 7 of 44 files at r2, 14 of 78 files at r3, 1 of 4 files at r4.
Review status: 23 of 99 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[atlasdb-api/src/main/java/com/palantir/atlasdb/spi/AtlasDbFactory.java, line 40 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku-zAnwwPY4tfX--4cX:-Ku-zAnwwPY4tfX--4cY:b300jk9) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-api/src/main/java/com/palantir/atlasdb/spi/AtlasDbFactory.java#L40)):*
> ```Java
> 
>     KeyValueService createRawKeyValueService(KeyValueServiceConfig config, Optional<LeaderConfig> leaderConfig,
>             boolean initializeAsync);
> ```

Note that this is a dev break, as our internal auth product has custom implementations of AtlasDbFactory.

---

*[atlasdb-cassandra/build.gradle, line 23 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku-zL28TkR8L_5OKp9e:-Ku-zL28TkR8L_5OKp9f:b-q4sh88) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-cassandra/build.gradle#L23)):*
> ```Gradle
> 
>   explicitShadow project(":atlasdb-client")
>   explicitShadow project(":atlasdb-processors")
> ```

Have you checked the jar produced is still right after this change?

---

*[atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolIntegrationTest.java, line 167 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku-ze9MrCFCUHDV7V6l:-Ku-ze9MrCFCUHDV7V6m:bquo255) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolIntegrationTest.java#L167)):*
> ```Java
>     @Test
>     public void testPoolGivenNoOptionTalksToBlacklistedHosts() {
>         clientPool.getBlacklistedHosts().putAll(
> ```

Is this method VisibleForTesting or actually used by production code?
It's beyond the scope of this PR, but `clientPool` should not be exposing its list of blacklisted hosts. Rather, it should expose methods `addBlacklistedHosts` and `clearBlacklistedHosts`.

---

*[atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceIntegrationTest.java, line 137 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku0-EeKyNynnxYkiw59:-Ku0-EeL8BlzLt3mVQdh:bf2nqiz) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceIntegrationTest.java#L137)):*
> ```Java
>     }
> 
>     // Suppressing Slf4j warnings for test
> ```

This comment isn't needed IMHO (same about 65 lines later).

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java, line 133 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku00O9ImBKCbLJNzl8t:-Ku00O9ImBKCbLJNzl8u:b-o5f1pe) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java#L133)):*
> ```Java
>     volatile RangeMap<LightweightOppToken, List<InetSocketAddress>> tokenMap = ImmutableRangeMap.of();
> 
>     private final CassandraKeyValueServiceConfig config;
> ```

The mix of final and non-final variables here is a bit messy (as is the fact that the fields are inserted randomly in between a couple of helper classes).

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java, line 154 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku052HqKnzpdOoLJQY0:-Ku052HqKnzpdOoLJQY1:bu7y1bm) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java#L154)):*
> ```Java
>         implements AsyncInitializer, CassandraKeyValueService {
> 
>     public static class InitializeCheckingWrapper implements AutoDelegate_CassandraKeyValueService {
> ```

Could this be made into a generic class, to avoid having to write it separately in each AsyncInitialized class, or does the AutoDelegate framework prevent this?

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java, line 266 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku05k4QJvD3XW4TZXr3:-Ku05W8csc6fYIXsxJnE:b-7t5sbg) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java#L266)):*
> ```Java
>     @Override
>     public synchronized void cleanUpOnInitFailure() {
>         // no-op
> ```

Can this happen after partial initialisation?

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraVerifier.java, line 215 at r3](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku-KjCtp7laHSUEIrZD-r3-215:-Ku05r_kjodUZ-1AtS9t:br51psk) ([raw file](https://github.com/palantir/atlasdb/blob/c8ce82064b401e5e042aa18ae6ebda5b1768bcc2/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraVerifier.java#L215)):*
<details><summary><i>Previously, fsamuel-bs (Samuel Souza) wrote…</i></summary><blockquote>

This allows a keyspace to be created when we have a 3 node Cassandra cluster and just two of them are up.

This is needed to allow our ETE tests to run (we drop and re-create the keyspace between runs). Partially fixes https://github.com/palantir/atlasdb/issues/1705.
</blockquote></details>

The first sentence of this could be added as a comment in the code.
The second sentence could be in the commit message :)

---

*[atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKvsWrapperTest.java, line 34 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku06F9l5zRFJTz5PZmE:-Ku06F9mWD76oCAqQ5zo:b-738mze) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKvsWrapperTest.java#L34)):*
> ```Java
> import com.palantir.exception.NotInitializedException;
> 
> public class CassandraKvsWrapperTest {
> ```

Would this be a good place to test that we `assertNotInitialized` when intializing?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2300)*
<!-- Sent from Reviewable.io -->

gsheasby commented:
Reviewed 2 of 10 files at r1, 8 of 44 files at r2, 28 of 78 files at r3, 2 of 4 files at r4.
Review status: 63 of 99 files reviewed at latest revision, 23 unresolved discussions, some commit checks failed.

---

*[atlasdb-commons/src/main/java/com/palantir/async/initializer/AsyncInitializer.java, line 59 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku07I5qQiyOCdQqjZvN:-Ku07I5qQiyOCdQqjZvO:bk48tcq) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-commons/src/main/java/com/palantir/async/initializer/AsyncInitializer.java#L59)):*
> ```Java
>             tryToInitializeIfNotInitialized();
>         } catch (Throwable throwable) {
>             log.warn("Failed to initialize {} in the first attempt, will initialize asynchronously.",
> ```

This should quite possibly be info, as it's expected to happen somewhere in the world where everything gets started up at once, and it's not something that impl needs to care about.

---

*[atlasdb-commons/src/main/java/com/palantir/async/initializer/AsyncInitializer.java, line 71 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku07RXBoaOi6Rl8gEXr:-Ku07RXBoaOi6Rl8gEXs:b-1vyrkn) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-commons/src/main/java/com/palantir/async/initializer/AsyncInitializer.java#L71)):*
> ```Java
>             try {
>                 tryToInitializeIfNotInitialized();
>                 log.warn("Initialized {} asynchronously.",
> ```

This should definitely be info!

---

*[atlasdb-commons/src/main/java/com/palantir/async/initializer/AsyncInitializer.java, line 83 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku07WciDKkE-J2Kpxpm:-Ku07WciDKkE-J2Kpxpn:b-gedknr) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-commons/src/main/java/com/palantir/async/initializer/AsyncInitializer.java#L83)):*
> ```Java
>     default void tryToInitializeIfNotInitialized() {
>         if (isInitialized()) {
>             log.warn("{} was initialized underneath us.",
> ```

I'm ok with this being a warning :)

---

*[atlasdb-commons/src/main/java/com/palantir/async/initializer/AsyncInitializer.java, line 101 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku07abNJDBG9F9pt99K:-Ku07abNJDBG9F9pt99L:bwxbdlb) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-commons/src/main/java/com/palantir/async/initializer/AsyncInitializer.java#L101)):*
> ```Java
>      * @throws IllegalStateException if object has already been initialized.
>      */
>     // TODO: Make this method protected abstract.
> ```

What's stopping us from just doing this? Java9 again?

---

*[atlasdb-commons/src/main/java/com/palantir/exception/NotInitializedException.java, line 21 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku07u3LopkOwtjsFNCv:-Ku07u3LopkOwtjsFNCw:bbic89n) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-commons/src/main/java/com/palantir/exception/NotInitializedException.java#L21)):*
> ```Java
> public class NotInitializedException extends RuntimeException {
>     public NotInitializedException(String objectNotInitialized) {
>         super(String.format("The %s is not initialized yet", objectNotInitialized));
> ```

Can we `SafeArg` this?

---

*[atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagersInitializer.java, line 68 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku09nFa385mgXUP0rXf:-Ku09nFa385mgXUP0rXg:b-d2ww1g) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagersInitializer.java#L68)):*
> ```Java
>         }
> 
>         LoggingArgs.hydrate(keyValueService.getMetadataForTables());
> ```

We've lost a TODO from previously, but nothing seems to have changed. Was it removed on purpose?

---

*[atlasdb-ete-tests/docker/conf/atlasdb-ete.multiple-cassandra-async.yml, line 19 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku0Dl5BB_SHfegR9KLM:-Ku0Dl5BB_SHfegR9KLN:bnd58wq) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-ete-tests/docker/conf/atlasdb-ete.multiple-cassandra-async.yml#L19)):*
> ```YAML
>       - cassandra2:9160
>       - cassandra3:9160
>     keyspace: atlasete
> ```

nit: use namespace instead of keyspace

---

*[atlasdb-ete-tests/src/main/java/com/palantir/atlasdb/AtlasDbEteServer.java, line 87 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku0EBYzmChQwcaRFKP2:-Ku0EBYzmChQwcaRFKP3:b4d74v9) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-ete-tests/src/main/java/com/palantir/atlasdb/AtlasDbEteServer.java#L87)):*
> ```Java
>     }
> 
>     private TransactionManager retryOnTransactionManagersCreationFailure(AtlasDbConfig config, Environment environment)
> ```

This could instead be called `createTransactionManagerWithRetry`.

---

*[atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/EteSetup.java, line 99 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku0EgKbgaIGi41vB6CE:-Ku0EgKbgaIGi41vB6CF:b-lcoosi) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/EteSetup.java#L99)):*
> ```Java
>     }
> 
>     static RuleChain setup(
> ```

This one can be private (and a few others in this class).

---

*[atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/TimeLockMigrationEteTest.java, line 166 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku0GASf_xd5KZBDP3_D:-Ku0GASf_xd5KZBDP3_E:bza33nq) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/TimeLockMigrationEteTest.java#L166)):*
> ```Java
>     }
> 
>     // Note that this check is a bit hacky, as it depends on finding a particular log message
> ```

...and one that doesn't shout out "transaction manager creation failure"! Can this be improved (perhaps in a future PR)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2300)*
<!-- Sent from Reviewable.io -->

nziebart commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/cleaner/KeyValueServicePuncherStore.java:
```diff
@@ -36,16 +37,61 @@
-public final class KeyValueServicePuncherStore implements PuncherStore {
+@AutoDelegate(typeToExtend = PuncherStore.class)
+public final class KeyValueServicePuncherStore implements PuncherStore, AsyncInitializer {
+    public static class InitializingWrapper implements AutoDelegate_PuncherStore {
```
what if we moved all init code into this class, and made it non-static? then it seems like you could make `AsyncInitializer` an abstract class

gsheasby commented:
Reviewed 1 of 10 files at r1, 5 of 44 files at r2, 29 of 78 files at r3, 1 of 4 files at r4.
Review status: all files reviewed at latest revision, 27 unresolved discussions, some commit checks failed.

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/cleaner/KeyValueServicePuncherStore.java, line 50 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ksxc6o4xt-CJ4KJt_rT-r4-50:-Ku4-ROQk70F1KD-dxou:b-tsolvh) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/cleaner/KeyValueServicePuncherStore.java#L50)):*
<details><summary><i>Previously, nziebart wrote…</i></summary><blockquote>

what if we moved all init code into this class, and made it non-static? then it seems like you could make `AsyncInitializer` an abstract class
</blockquote></details>

+1, that could cut down on a lot of boilerplate and make adding new `AsyncInitializer` less prone to human error.

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/cleaner/KeyValueServiceScrubberStore.java, line 60 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku40LpN28boLU9eiJjT:-Ku40LpN28boLU9eiJjU:b-ki0njb) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/cleaner/KeyValueServiceScrubberStore.java#L60)):*
> ```Java
>  *
>  *
>  * @author ejin
> ```

BTW, we can proactively remove `@author` tags when we see them. (this is very much optional)

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/LockStoreImpl.java, line 84 at r1](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ksxc6o4xt-CJ4KJt_rU-r1-84:-Ku44FUrsZVIACZfyCOo:b3y1s3j) ([raw file](https://github.com/palantir/atlasdb/blob/bf5273620b7a223aa8705d3db17fcb187844d19e/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/LockStoreImpl.java#L84)):*
<details><summary><i>Previously, fsamuel-bs (Samuel Souza) wrote…</i></summary><blockquote>

Agreed that the current logic makes more sense. We should probably check that this won't break the internal large product though. Maybe running the tests with a local patch would be enough?
</blockquote></details>

Does the large internal product have any additional protections around creating a TM? I wouldn't be surprised if it just inherited any protections from us, i.e. the semantic change here would be invisible to them. Worth doing some code inspection, though.

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/CellsSweeper.java, line 58 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku44k1fGo3aR01wpmB_:-Ku44k1fGo3aR01wpmBa:bho14p4) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/CellsSweeper.java#L58)):*
> ```Java
>             KeyValueService keyValueService,
>             Collection<Follower> followers) {
>         this(txManager, keyValueService, followers, AtlasDbConstants.DEFAULT_INITIALIZE_ASYNC);
> ```

I'm not a fan of the constructor proliferation here. This is not a class that should be used outside of AtlasDB; indeed, it should only be used in a small handful of places inside AtlasDB. Can we reduce the number of constructors?

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SerializableTransactionManager.java, line 52 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku45t4Ick3BUDNQBojF:-Ku45t4Ick3BUDNQBojG:bcx3yj) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SerializableTransactionManager.java#L52)):*
> ```Java
>                 try {
>                     // SerializableTransactionManager should throw until the underlying KVS is initialized.
>                     // TODO(ssouza): replace with KVS healthcheck status when that gets implemented.
> ```

Are we planning to implement that? (perhaps I missed some discussion at some point)

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SerializableTransactionManager.java, line 78 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku46RWumeH6EiYlMsVv:-Ku46RWumeH6EiYlMsVw:b32vjw) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SerializableTransactionManager.java#L78)):*
> ```Java
>      */
>     protected SerializableTransactionManager() {
>         this(null, null, null, null, null, null, null, null, null);
> ```

<3

---

*[atlasdb-jdbc/src/main/java/com/palantir/atlasdb/keyvalue/jdbc/JdbcAtlasDbFactory.java, line 39 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku46hVNalw-6wsypbpL:-Ku46hVNalw-6wsypbpM:b-rqq1sx) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-jdbc/src/main/java/com/palantir/atlasdb/keyvalue/jdbc/JdbcAtlasDbFactory.java#L39)):*
> ```Java
>     @Override
>     public KeyValueService createRawKeyValueService(KeyValueServiceConfig config, Optional<LeaderConfig> leaderConfig,
>             boolean initializeAsync) {
> ```

Might want to note that we ignore async initialisation.
We should probably also warn if this is somehow called with async=true.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2300)*
<!-- Sent from Reviewable.io -->

gmaretic commented:
Review status: 83 of 101 files reviewed at latest revision, 27 unresolved discussions, some commit checks failed.

---

*[atlasdb-api/src/main/java/com/palantir/atlasdb/spi/AtlasDbFactory.java, line 40 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku-zAnwwPY4tfX--4cX:-Ku5CjyOdZzbT231jWiG:bg5ta71) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-api/src/main/java/com/palantir/atlasdb/spi/AtlasDbFactory.java#L40)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary><blockquote>

Note that this is a dev break, as our internal auth product has custom implementations of AtlasDbFactory.
</blockquote></details>

Good flag, we need to make sure to add that to the release notes.

---

*[atlasdb-cassandra/build.gradle, line 23 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku-zL28TkR8L_5OKp9e:-Ku5D1v_8r56uUjC8o7n:bumedaj) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-cassandra/build.gradle#L23)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary><blockquote>

Have you checked the jar produced is still right after this change?
</blockquote></details>

The benchmarking test would fail if the jar wasn't good :)

---

*[atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolIntegrationTest.java, line 167 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku-ze9MrCFCUHDV7V6l:-Ku5DKPU7nAb9dhAQfjF:b-8ys0ce) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolIntegrationTest.java#L167)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary><blockquote>

Is this method VisibleForTesting or actually used by production code?
It's beyond the scope of this PR, but `clientPool` should not be exposing its list of blacklisted hosts. Rather, it should expose methods `addBlacklistedHosts` and `clearBlacklistedHosts`.
</blockquote></details>

Good spot, this was done so that we can use the interface in tests, but we ended up using the Impl, anyway, so I reverted it.

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java, line 133 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku00O9ImBKCbLJNzl8t:-Ku5Ob3KhKecloQEsiY8:b-cyncct) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java#L133)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary><blockquote>

The mix of final and non-final variables here is a bit messy (as is the fact that the fields are inserted randomly in between a couple of helper classes).
</blockquote></details>

Moved things around to make it more legible

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java, line 154 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku052HqKnzpdOoLJQY0:-Ku5OjgzSRX8-QsQa9AF:buszw5c) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java#L154)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary><blockquote>

Could this be made into a generic class, to avoid having to write it separately in each AsyncInitialized class, or does the AutoDelegate framework prevent this?
</blockquote></details>

We could have a processor to autogenerate the code, but that is quite a bit of work, so was backlogged.

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java, line 266 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku05k4QJvD3XW4TZXr3:-Ku5P0-aWtTpvB6lzUqI:bkkrx3r) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java#L266)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary><blockquote>

Can this happen after partial initialisation?
</blockquote></details>

In general, it is really hard to enforce thread safety on the interface level. As long as the users are not trying to break things, this should never be called if there is an initialization in progress (there is a fix in part 4 that actually ensures this on the interface level).

---

*[atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraVerifier.java, line 215 at r3](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku-KjCtp7laHSUEIrZD-r3-215:-Ku5S-xuLqYHtiqwBlQL:b-66g9zj) ([raw file](https://github.com/palantir/atlasdb/blob/c8ce82064b401e5e042aa18ae6ebda5b1768bcc2/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraVerifier.java#L215)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary><blockquote>

The first sentence of this could be added as a comment in the code.
The second sentence could be in the commit message :)
</blockquote></details>

I'm actually not sure if it's needed for the tests anymore, but definitely an improvement so leaving it in.

---

*[atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKvsWrapperTest.java, line 34 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku06F9l5zRFJTz5PZmE:-Ku5UO_Iq-l_XQMcuRPN:bubg79f) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKvsWrapperTest.java#L34)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary><blockquote>

Would this be a good place to test that we `assertNotInitialized` when intializing?
</blockquote></details>

I think Mockito does not interact nicely with default methods unfortunately.

---

*[atlasdb-commons/src/main/java/com/palantir/async/initializer/AsyncInitializer.java, line 101 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku07abNJDBG9F9pt99K:-Ku5V7Y7zPVrHYC1TGQC:boyp5uj) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-commons/src/main/java/com/palantir/async/initializer/AsyncInitializer.java#L101)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary><blockquote>

What's stopping us from just doing this? Java9 again?
</blockquote></details>

It's not a feature in Java 8

---

*[atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagersInitializer.java, line 68 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku09nFa385mgXUP0rXf:-Ku5XFLJW2xbwfrz8wux:b-kwjy1l) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagersInitializer.java#L68)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary><blockquote>

We've lost a TODO from previously, but nothing seems to have changed. Was it removed on purpose?
</blockquote></details>

What was the todo? I don't see it in the history

---

*[atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/TimeLockMigrationEteTest.java, line 166 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku0GASf_xd5KZBDP3_D:-Ku5YYLY4tIYnQGEWsRV:bjw3j9u) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/TimeLockMigrationEteTest.java#L166)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary><blockquote>

...and one that doesn't shout out "transaction manager creation failure"! Can this be improved (perhaps in a future PR)?
</blockquote></details>

I don't think there is much we can do about the test, but I'm all for throwing a better exception 

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/cleaner/KeyValueServicePuncherStore.java, line 50 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ksxc6o4xt-CJ4KJt_rT-r4-50:-Ku5ZkIxlFfa5HUm3NE8:bnvxmvu) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/cleaner/KeyValueServicePuncherStore.java#L50)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary><blockquote>

+1, that could cut down on a lot of boilerplate and make adding new `AsyncInitializer` less prone to human error.
</blockquote></details>

Ah, you mean so that the main class does not implement/extend AsyncInitializer? We can try it out, but in a separate PR

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/LockStoreImpl.java, line 84 at r1](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ksxc6o4xt-CJ4KJt_rU-r1-84:-Ku5a08DcZ9xXq2WTu0Q:b-utfvf4) ([raw file](https://github.com/palantir/atlasdb/blob/bf5273620b7a223aa8705d3db17fcb187844d19e/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/LockStoreImpl.java#L84)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary><blockquote>

Does the large internal product have any additional protections around creating a TM? I wouldn't be surprised if it just inherited any protections from us, i.e. the semantic change here would be invisible to them. Worth doing some code inspection, though.
</blockquote></details>

Let's reach out to them to make sure.

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SerializableTransactionManager.java, line 52 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku45t4Ick3BUDNQBojF:-Ku5aXn4kd0XEhIzZjcs:b-5n3rq6) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SerializableTransactionManager.java#L52)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary><blockquote>

Are we planning to implement that? (perhaps I missed some discussion at some point)
</blockquote></details>

Not in the scope of this epic

---

*[atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SerializableTransactionManager.java, line 78 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku46RWumeH6EiYlMsVv:-Ku5ad45FSj7eQEw5ah4:b-yra7ev) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SerializableTransactionManager.java#L78)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary><blockquote>

<3
</blockquote></details>

My favourite :dancer: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2300)*
<!-- Sent from Reviewable.io -->

gsheasby commented:
Reviewed 1 of 78 files at r3, 18 of 18 files at r5, 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[atlasdb-commons/src/test/java/com/palantir/async/initializer/AsyncInitializerTest.java, line 90 at r6](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-KuK-SjNjpP3QbXprCjR:-KuK-SjNjpP3QbXprCjS:b-mci70h) ([raw file](https://github.com/palantir/atlasdb/blob/10d632fa3081617cfe9258979bf0f6b983ea2164/atlasdb-commons/src/test/java/com/palantir/async/initializer/AsyncInitializerTest.java#L90)):*
> ```Java
>         assertThat(initializer.initializationAttempts).isEqualTo(0);
>     }
> }
> ```

For consistency, we need a newline at EOF. This would be a checkstyle thing, but atlasdb-commons isn't checked.

---

*[atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagersInitializer.java, line 68 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku09nFa385mgXUP0rXf:-KuK1KFpi5qlR0RQ3lah:b-vs6ekv) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagersInitializer.java#L68)):*
<details><summary><i>Previously, gmaretic wrote…</i></summary><blockquote>

What was the todo? I don't see it in the history
</blockquote></details>

The code used to live in TransactionManagers. The TODO was:
```
// Prime the key value service with logging information.
// TODO (jkong): Needs to be changed if/when we support dynamic table creation.
LoggingArgs.hydrate(kvs.getMetadataForTables());
```

---

*[atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/TimeLockMigrationEteTest.java, line 166 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku0GASf_xd5KZBDP3_D:-KuK1_79Y8JkOzRwx9g5:b-7o9plw) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/TimeLockMigrationEteTest.java#L166)):*
<details><summary><i>Previously, gmaretic wrote…</i></summary><blockquote>

I don't think there is much we can do about the test, but I'm all for throwing a better exception 
</blockquote></details>

In this PR or elsewhere?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2300)*
<!-- Sent from Reviewable.io -->

gmaretic commented:
*[atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagersInitializer.java, line 68 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku09nFa385mgXUP0rXf:-KuKQW74Q_TJkXX-bXm3:bbztuuz) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagersInitializer.java#L68)):*
> // Prime the key value service with logging information.
> // TODO (jkong): Needs to be changed if/when we support dynamic table creation.
Readded the TODO

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2300)*
<!-- Sent from Reviewable.io -->

gmaretic commented:
*[atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/TimeLockMigrationEteTest.java, line 166 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku0GASf_xd5KZBDP3_D:-KuKQhYBs3n4UctwTETd:bla298) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/TimeLockMigrationEteTest.java#L166)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary><blockquote>

In this PR or elsewhere?
</blockquote></details>

This is completely out of the scope of this PR

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2300)*
<!-- Sent from Reviewable.io -->

gsheasby commented:
Reviewed 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[atlasdb-api/src/main/java/com/palantir/atlasdb/spi/AtlasDbFactory.java, line 40 at r4](https://reviewable.io:443/reviews/palantir/atlasdb/2300#-Ku-zAnwwPY4tfX--4cX:-KuKdXsuyzJJE6mZk0kK:btybvcl) ([raw file](https://github.com/palantir/atlasdb/blob/f961e01ceeaff4168576936db55660115dee2a70/atlasdb-api/src/main/java/com/palantir/atlasdb/spi/AtlasDbFactory.java#L40)):*
<details><summary><i>Previously, gmaretic wrote…</i></summary><blockquote>

Good flag, we need to make sure to add that to the release notes.
</blockquote></details>

:+1: will be handled in a separate PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/2300)*
<!-- Sent from Reviewable.io -->
